.travis.yml

```diff
@@ -1,14 +1,29 @@
-language: python
+dist: xenial
 
+language: python
 python:
-  - "3.4"
-  - "3.5"
-  - "3.6"
+  - 3.4
+  - 3.5
+  - 3.6
+  - 3.7
+
+env:
+  matrix:
+    - LANG=c
+    - LANG=go
+    - LANG=java
+    - LANG=javascript
+    - LANG=powershell
+    - LANG=python
```

Comment:
Refer to https://docs.travis-ci.com/user/customizing-the-build/#build-timeouts and the failed latest commit due to exceeding 50 minutes.

I guess with the growths of number of supported languages this change was inevitable. With the growths of number of supported algorithms there will be need to parallelize across them as well. However, it seems that parallelization across languages is enough for now.

Reply:
Wow, this looks really cool!

One thing that bothers me is that we will run exactly same set of unit tests for all of these runs. While it might seem insignificant, I would still prefer for every run to be isolated.

Would it make sense to add another row here specifically to run unit tests? However we should keep in mind that only unit tests report code coverage (this is by design), and so only this single run would report it. I'm not sure about travis specifics and how it would affect the calculation of the result code coverage.

Reply:
For some reason powershell takes much more time to run: <10 minutes against 40:
<img width="1021" alt="Screen Shot 2019-10-25 at 6 09 03 PM" src="https://user-images.githubusercontent.com/457589/67611898-a50d4500-f752-11e9-9554-dfb8ee7eae7c.png">

Reply:
> I'm not sure about travis specifics and how it would affect the calculation of the result code coverage.

Not sure, but I think that only the last one for PR/commit is taken into account.



Ah! Even with breakage in terms of languages one job hitted 50 minutes!


Hm, what do you think about moving to [Azure Pipelines](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/phases?view=azure-devops&tabs=yaml#timeouts)? I can create a minimal config here or in a separate PR. But you will have to sign up there.

Reply:
> Not sure, but I think that only the last one for PR/commit is taken into account.

What I mean is that even for one PR/commit there are many builds, each potentially reporting different coverage. So if we separate unit and e2e tests, only subset of our builds will report coverage, and somehow they will be aggregated into single metric.

> what do you think about moving to Azure Pipelines?

I don't have a strong argument against, however I think it's worth checking what exactly makes powershell tests runnning so long as sometimes we need an ability to run e2e tests locally.

From pure technical point it could be one of the following:
- installing dependencies;
- code generation;
- execution.

Or some combination of those, of course :)

Reply:
> So if we separate unit and e2e tests, only subset of our builds will report coverage, and somehow they will be aggregated into single metric.

You can click here and take a look. Seems that a coverage tool deals with this situation fine:

![image](https://user-images.githubusercontent.com/25141164/67613556-74132b80-f7b7-11e9-8d87-d80073114dca.png)


> I don't have a strong argument against …

Fine! Then I'd like to prepare a separate PR for this today.

> ... however I think it's worth checking what exactly makes powershell tests running so long as sometimes we need an ability to run e2e tests locally.

Installing `PowerShell Core` takes 13 seconds, so to be honest, I simply don't know, sorry 🙁 . I only can confirm that the same situation is in my local Windows environment with native PowerShell (not Core): awful timings.


![image](https://user-images.githubusercontent.com/25141164/67613593-e421b180-f7b7-11e9-88cb-1d58c10e9a9e.png)

Reply:
> Then I'd like to prepare a separate PR for this today.

Can you please register at Azure Pipelines and push an empty `azure-pipelines.yml` file in `master`, so that it will be possible to see a result for my incoming PR? Otherwise, with that file in `master` builds will not be triggered.

Reply:
@krinart Please see #116.